src/ray/common/scheduling/cluster_resource_scheduler.cc

```diff
@@ -19,6 +50,87 @@ std::string NodeResources::DebugString() {
   return buffer.str();
 }
 
+std::string PrintVector(std::vector<double> &vector) {
+  std::stringstream buffer;
+
+  buffer << "[";
+  for (size_t i = 0; i < vector.size(); i++) {
+    buffer << vector[i];
+    if (i < vector.size() - 1) {
+      buffer << ", ";
+    }
+  }
+  buffer << "]";
+  return buffer.str();
+}
+
+bool NodeResourceInstances::operator==(const NodeResourceInstances &other) {
+  for (size_t i = 0; i < PredefinedResources_MAX; i++) {
+    if (!EqualVectors(this->predefined_resources[i].total,
+                      other.predefined_resources[i].total)) {
+      return false;
+    }
+    if (!EqualVectors(this->predefined_resources[i].available,
+                      other.predefined_resources[i].available)) {
+      return false;
+    }
+  }
+
+  if (this->custom_resources.size() != other.custom_resources.size()) {
+    return true;
+  }
+
+  for (auto it1 = this->custom_resources.begin(); it1 != this->custom_resources.end();
+       ++it1) {
+    auto it2 = other.custom_resources.find(it1->first);
+    if (it2 == other.custom_resources.end()) {
+      return false;
+    }
+    if (!EqualVectors(it1->second.total, it2->second.total)) {
+      return false;
+    }
+    if (!EqualVectors(it1->second.available, it2->second.available)) {
+      return false;
+    }
+  }
+  return true;
+}
+
+std::string NodeResourceInstances::DebugString() {
+  std::stringstream buffer;
+  buffer << "  node predefined resources {";
+  for (size_t i = 0; i < this->predefined_resources.size(); i++) {
+    buffer << "(" << PrintVector(predefined_resources[i].total) << ":"
+           << PrintVector(this->predefined_resources[i].available) << ") ";
+  }
+  buffer << "}" << std::endl;
+
+  buffer << "  node custom resources {";
+  for (auto it = this->custom_resources.begin(); it != this->custom_resources.end();
+       ++it) {
+    buffer << it->first << ":(" << PrintVector(it->second.total) << ":"
+           << PrintVector(it->second.available) << ") ";
+  }
+  buffer << "}" << std::endl;
+  return buffer.str();
+};
+
+TaskResourceInstances NodeResourceInstances::GetAvailableResourceInstances() {
+  TaskResourceInstances task_resources;
+  task_resources.predefined_resources.resize(PredefinedResources_MAX);
+
+  for (size_t i = 0; i < this->predefined_resources.size(); i++) {
+    task_resources.predefined_resources[i] = this->predefined_resources[i].available;
+  }
+
+  for (auto it = this->custom_resources.begin(); it != this->custom_resources.end();
```

Comment:
Prefer to use `for (const auto it : this->custom_resources) {`

Reply:
fixed